src/pages/game/memory.tsx

```diff
@@ -1,7 +1,184 @@
 import type { NextPage } from 'next';
+import { useState, useEffect } from 'react';
+import { Box, Button, Grid, Progress, Text } from '@chakra-ui/react';
+import MemoryGameBoxBtn from '../../components/MemoryGameBoxBtn';
+import useStage, { IStageHookProps } from '../../hooks/useStage';
+export const GRID_ITEM_COUNT = [
+  { size: 2, count: 4 },
+  {
+    size: 2,
+    count: 1,
+  },
+  {
+    size: 2,
+    count: 2,
+  },
+  {
+    size: 2,
+    count: 3,
+  },
+  {
+    size: 3,
+    count: 2,
+  },
+  {
+    size: 3,
+    count: 3,
+  },
+  {
+    size: 3,
+    count: 4,
+  },
+  {
+    size: 4,
+    count: 3,
+  },
+  {
+    size: 4,
+    count: 4,
+  },
+  {
+    size: 4,
+    count: 5,
+  },
+  {
+    size: 5,
+    count: 4,
+  },
+  {
+    size: 5,
+    count: 5,
+  },
+  {
+    size: 5,
+    count: 6,
+  },
+  {
+    size: 5,
+    count: 7,
+  },
+  {
+    size: 5,
+    count: 8,
+  },
+  {
+    size: 5,
+    count: 9,
+  },
+];
 
 const Memory: NextPage = () => {
-  return <div>Memory</div>;
+  const [correctIndexs, setCorrectIndexs] = useState<number[]>([]);
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+  const [clickCount, setClickCount] = useState(0);
+  // const [life, setLife] = useState(3);
+  const { stage, nextStage, getCorrectIndexs }: IStageHookProps = useStage();
+
+  const viewBtn = () => {
+    console.log('correct:', correctIndexs);
+    return Array.from({ length: GRID_ITEM_COUNT[stage].size * GRID_ITEM_COUNT[stage].size }).map(
+      (_, idx) => {
+        let isReload = correctIndexs.length === GRID_ITEM_COUNT[stage].count;
+        let count = isReload ? correctIndexs.findIndex((e: number) => e === idx) : -1;
+        return (
+          <MemoryGameBoxBtn
+            changedColor={correctIndexs[clickCount] === idx ? 'green.600' : 'red.600'}
+            key={`${idx}-grid-item`}
+            count={count}
+            isLoading={isLoading}
+            setIsLoading={setIsLoading}
+            stage={stage}
+            setClickCount={setClickCount}
+            clickCount={clickCount}
+          />
+        );
+      },
+    );
+  };
+
+  useEffect(() => {
+    setTimeout(() => {
+      if (clickCount === 0) setCorrectIndexs(getCorrectIndexs());
+    }, 900);
+  }, [getCorrectIndexs, clickCount]);
+
+  useEffect(() => {
+    if (clickCount === GRID_ITEM_COUNT[stage].count || clickCount === -1) {
+      setIsLoading(true);
+      setTimeout(() => {
+        if (clickCount !== -1) {
+          nextStage();
+        }
+        setClickCount(0);
+      }, 700);
+    }
+  }, [nextStage, clickCount, stage]);
```

Comment:
```suggestion
  const WRONG_TRACE = -1;
  const MEMORY_GAME_CHANGE_STAGE_TERM = 700;

  useEffect(() => {
    const isClickCorrectTrace = clickCount === GRID_ITEM_COUNT[stage].count;
    const isClickWrongTrace = clickCount === WRONG_TRACE;
    if (isClickCorrectTrace || isClickWrongTrace) {
      setIsLoading(true);
      setTimeout(() => {
        if (isClickWrongTrace) nextStage();
        setClickCount(0);
      }, MEMORY_GAME_CHANGE_STAGE_TERM);
    }
  }, [nextStage, clickCount, stage]);
```

만약 가독성을 높인다면 이런식으로 작성할 수 있을 것 같네요?
보통 `boolean` 값은 `is` prefix(접두사)를 붙여서 사용합니당